scipy/interpolate/src/__fitpack.cc

```diff
@@ -0,0 +1,425 @@
+#include <cstring>
+#include "__fitpack.h"
+
+namespace fitpack{
+
+/*
+ * B-spline evaluation routine.
+ */
+void
+_deBoor_D(const double *t, double x, int k, int ell, int m, double *result) {
+    /*
+     * On completion the result array stores
+     * the k+1 non-zero values of beta^(m)_i,k(x):  for i=ell, ell-1, ell-2, ell-k.
+     * Where t[ell] <= x < t[ell+1].
+     */
+    /*
+     * Implements a recursive algorithm similar to the original algorithm of
+     * deBoor.
+     */
+    double *hh = result + k + 1;
+    double *h = result;
+    double xb, xa, w;
+    int ind, j, n;
+
+    /*
+     * Perform k-m "standard" deBoor iterations
+     * so that h contains the k+1 non-zero values of beta_{ell,k-m}(x)
+     * needed to calculate the remaining derivatives.
+     */
+    result[0] = 1.0;
+    for (j = 1; j <= k - m; j++) {
+        memcpy(hh, h, j*sizeof(double));
+        h[0] = 0.0;
+        for (n = 1; n <= j; n++) {
+            ind = ell + n;
+            xb = t[ind];
+            xa = t[ind - j];
+            if (xb == xa) {
+                h[n] = 0.0;
+                continue;
+            }
+            w = hh[n - 1]/(xb - xa);
+            h[n - 1] += w*(xb - x);
+            h[n] = w*(x - xa);
+        }
+    }
+
+    /*
+     * Now do m "derivative" recursions
+     * to convert the values of beta into the mth derivative
+     */
+    for (j = k - m + 1; j <= k; j++) {
+        memcpy(hh, h, j*sizeof(double));
+        h[0] = 0.0;
+        for (n = 1; n <= j; n++) {
+            ind = ell + n;
+            xb = t[ind];
+            xa = t[ind - j];
+            if (xb == xa) {
+                h[m] = 0.0;
+                continue;
+            }
+            w = j*hh[n - 1]/(xb - xa);
+            h[n - 1] -= w;
+            h[n] = w;
+        }
+    }
+}
+
+
+/*
+ *  Find an interval such that t[interval] <= xval < t[interval+1].
+ */
+ssize_t
+_find_interval(const double* tptr, ssize_t len_t,
+               int k,
+               double xval,
+               ssize_t prev_l,
+               int extrapolate)
+{
+    ConstRealArray1D t = ConstRealArray1D(tptr, len_t);
+
+    ssize_t n = t.nelem - k - 1;
+    double tb = t(k);
+    double te = t(n);
+
+    if (xval != xval) {
+        // nan
+        return -1;
+    }
+
+    if (((xval < tb) || (xval > te)) && !extrapolate) {
+        return -1;
+    }
+    ssize_t l = (k < prev_l) && (prev_l < n) ? prev_l : k;
+
+    // xval is in support, search for interval s.t. t[interval] <= xval < t[l+1]
+    while ((xval < t(l)) && (l != k)) {
+        l -= 1;
+    }
+
+    l += 1;
+    while ((xval >= t(l)) && (l != n)) {
+        l += 1;
+    }
+
+    return l-1;
+}
```

Comment:
IIRC (been a while) there are subtleties on how a binary search handles runs of equal values. Here it would be multiple knots, which are possible at the beginning, at the end and in the middle of the array. I guess it'll need to be carefully tested.

Reply:
That sounds reasonable. If it helps, it might also be worth looking at ``std::upper_bound``